Options / Environment / Config Parsing

StevenXL opened the pull request:
Syd - I'm opening up a PR for discussion so that we can implement proper option parsing into the `hastory` project.

The first thing I've done here is to change the [genGatherWrapperScript](https://github.com/StevenXL/hastory/blob/3b9d87205354d41c50d517a912b5b8aa3f1d8759/hastory-cli/src/Hastory/Cli/Commands/GenGatherWrapper.hs#L11) so that if the remote server information is available, then the generated script will include that data in the command that is outputted. This functionality is [tested here](https://github.com/StevenXL/hastory/blob/3b9d87205354d41c50d517a912b5b8aa3f1d8759/hastory-cli/test/Hastory/Cli/Commands/GenGatherWrapperSpec.hs#L1). (If you squint, the `beforeAll` hook get us really close to a property test, but I don't think it is worth the time to write a proper `GenValid` instance for `BaseURL`).

I've read your post on [Option Parsing in Haskell - Part 2](https://cs-syd.eu/posts/2020-05-14-option-parsing-2) and found it to be very good. I'm looking forward to putting that into practice.

I have three questions that would be really helpful in getting me started here.

First, my understanding is that the data constructors for the `Command` data type have payloads that are specific to that command. Another way to phrase this is that any data that would be used by more than 1 command belongs in the `Flag` data type? In our specific case, the remote server information can be used by [gater](https://github.com/StevenXL/hastory/blob/3b9d87205354d41c50d517a912b5b8aa3f1d8759/hastory-cli/src/Hastory/Cli.hs#L25) command and the [genGatherWrapperScript](https://github.com/StevenXL/hastory/blob/3b9d87205354d41c50d517a912b5b8aa3f1d8759/hastory-cli/src/Hastory/Cli.hs#L26) command, and thus that data belongs in the [Flag](https://github.com/StevenXL/hastory/blob/3b9d87205354d41c50d517a912b5b8aa3f1d8759/hastory-cli/src/Hastory/Cli/OptParse/Types.hs#L31). (I'm pretty sure the remote server info belongs in the `Flag` data type, but I want to make sure my reasoning for *why* is correct - there's a term for reaching the right conclusion for the wrong reasons). 

The second question I have is whether it is OK to redefine the `Flag`'s `flagStorageServer` field from `Maybe FilePath` to `Maybe BaseUrl`. In your post, you mention that it is best to "Use simple types", but later on you also mention that "...we are allowed to pre-process a FilePath into a Path Abs Dir, but not list the directory to get a [Path Abs File]. The first part is considered parsing. The second part is considered application logic." I have a very strong preference to parse the data into its domain type as early on as possible. The downside to doing this "within" `optparse-applicative` is that we lose control of what the error looks like. The other option is to `die` in [combineToInstructions](https://github.com/StevenXL/hastory/blob/3b9d87205354d41c50d517a912b5b8aa3f1d8759/hastory-cli/src/Hastory/Cli/OptParse.hs#L27), which gives us greater control over the error message. I'm also just realizing now that `die`ing in `combineToInstructions` allows the data to be invalid in one source (say, the command line) but be valid in another, such as the environment. I'm actually conflicted about it now.

I had a third question, but it escaping me now.

StevenXL commented:
Now that I think about it, if a user passes in a malformed URL in the command line, but has a correct one in the environmental variables, we should err on the side of caution, assume that the user wanted the command line arguments to take precedence, and fail at the command-line argument parsing stage. So the con of converting the the `flagStorageServer` from `Maybe Text` to `Maybe BaseUrl` is that we lose control over the error message, but I feel better about that.

NorfairKing commented:
> Another way to phrase this is that any data that would be used by more than 1 command belongs in the Flag data type?

Not really. Any data that is necessary in ALL commands should be in the `Flags` data type. If some data belongs in multiple commands, just put it into both of those commands.

Contrived example:

```
data Command = CommandGather BaseUrl | CommandGenWrapperScript BaseUrl | CommandAliases
```

However, usually it makes sense to make a `GatherFlags` data type for a `gather` command. One for each command makes sense.

> The second question I have is whether it is OK to redefine the Flag's flagStorageServer field from Maybe FilePath to Maybe BaseUrl

For parsing flags, the question that matters is: "Do you need IO to do it?" If not, use `maybeReader` or `eitherReader` from `optparse-applicative`. If you need IO, then do the parsing in the `combineToInstructions` function.

Does this make sense?

StevenXL commented:
> > Another way to phrase this is that any data that would be used by more than 1 command belongs in the Flag data type?
> 
> Not really. Any data that is necessary in ALL commands should be in the `Flags` data type. If some data belongs in multiple commands, just put it into both of those commands.
> 
> Contrived example:
> 
> ```
> data Command = CommandGather BaseUrl | CommandGenWrapperScript BaseUrl | CommandAliases
> ```
> 
> However, usually it makes sense to make a `GatherFlags` data type for a `gather` command. One for each command makes sense.
> 
> > The second question I have is whether it is OK to redefine the Flag's flagStorageServer field from Maybe FilePath to Maybe BaseUrl
> 
> For parsing flags, the question that matters is: "Do you need IO to do it?" If not, use `maybeReader` or `eitherReader` from `optparse-applicative`. If you need IO, then do the parsing in the `combineToInstructions` function.
> 
> Does this make sense?

Yes it does. Thank you!

StevenXL commented:
@NorfairKing I'm working my way through this PR, and I wanted to give you the mental model that I have in regards to it. The way I am thinking about this is that I have a matrix in my head:

|               | Gather | GenGatherWrapperScript | ListRecentDirs | ChangeDir | GenChangeWrapperScript | SuggestAlias |
|---------------|--------|------------------------|----------------|-----------|------------------------|--------------|
| Command Line  |        |                        |                |           |                        |              |
| Environment   |        |                        |                |           |                        |              |
| Configuration |        |                        |                |           |                        |              |

Essentially, the user can run 6 commands. Each of these commands can be configured via 3 "modes" (what's the correct word here?). So I am just simply going to run down each command / mode combo.

NorfairKing commented:
> Each of these commands can be configured via 3 "modes" (what's the correct word here?). 

In three different _ways_ ?

> I think a great place to provide feedback will be after the GenGatherWrapperScript column is complete,

Yes, please ping me then.

StevenXL commented:
@NorfairKing Syd - obviously, no urgency on this, but I wanted to let you know that the PR is ready for some feedback. The end goal was to ensure that the `GenChangeWrapperScript` command could be configured via the command line, environment, and a configuration file.

NorfairKing commented on hastory-cli/src/Hastory/Cli/OptParse.hs:
```diff
@@ -1,62 +1,128 @@
+      if fileExists
+        then readConfigFile defaultConfigFile >>= displayingSchema
+        else pure defaultConfiguration
+    displayingSchema :: Maybe Configuration -> IO Configuration
```
using http://hackage.haskell.org/package/yamlparse-applicative-0.1.0.1/docs/YamlParse-Applicative-IO.html#v:readConfigFile
does this all for you, the failing if something goes wrong, the rendering the error etc.
You should not have a default configuration by the way, just return a `Maybe Configuration` from `getConfiguration`.

StevenXL replied:
Thanks Syd. I've removed the idea of a default configuration.

NorfairKing commented on hastory-cli/src/Hastory/Cli/OptParse.hs:
```diff
@@ -1,62 +1,128 @@
+envParser :: Env.Parser Env.Error Environment
+envParser =
+  Env.prefixed
+    "HASTORY_"
```
It looks like you use this prefix twice, please check by putting the help text in the output of `--help` (see the way the `intray-cli` does this).

StevenXL replied:
Indeed I did. Thank you.

NorfairKing commented on hastory-cli/src/Hastory/Cli/OptParse/Types.hs:
```diff
@@ -14,59 +20,160 @@ data Instructions =
+instance YamlSchema Configuration where
+  yamlSchema = parseEmptyConfigurationFile <|> parseObject
+    where
+      parseEmptyConfigurationFile =
```
I don't think you need to be able to deal with an empty configuratino file, I'd suggest removing this.

StevenXL replied:
I think I understand. [readConfigFile](https://hackage.haskell.org/package/yamlparse-applicative-0.1.0.1/docs/YamlParse-Applicative.html#v:readConfigFile) will not try to parse an empty file (or missing) file, so there is no reason for us to handle that case.

NorfairKing commented on hastory-cli/src/Hastory/Cli/OptParse/Types.hs:
```diff
@@ -14,59 +20,160 @@ data Instructions =
   deriving (Show, Eq)
 
+defaultConfiguration :: Configuration
+defaultConfiguration =
```
You _shouldn't_ need this. Best to remove it and fix what breaks.

StevenXL replied:
Done - I left these definitions in the spec file only for help with testing.

NorfairKing commented on hastory-cli/test/Hastory/Cli/OptParseSpec.hs:
```diff
@@ -0,0 +1,393 @@
+            DispatchGenGatherWrapperScript
+              GenGatherWrapperScriptSettings {genGatherWrapperScriptSetRemoteInfo = Just remoteInfo}
+
+emptyFlags :: Flags
```
You also shouldn't need empty flags or an empty configuration. Best to remove them and see what breaks.

StevenXL replied:
Is it OK to leave these in for testing only? Otherwise, I will need to build values over and over again with only one field a `Just` and the rest `Nothing`s. I'm perfectly fine w/ taking them out if you still want me to do so, but just wanted to point out that they are only in the spec files. Please let me know.

NorfairKing replied:
Ah yes if it's only used in the tests then that's fine. Thanks for pointing that out!

StevenXL commented:
@NorfairKing Hi Syd - got a few minutes free from work. Addressed most of the comments so far except for 1. Please let me know how you'd like to proceed. As discussed before, once we get this one command squared away, I'll apply the pattern to the rest. Thank you!

NorfairKing commented on hastory-cli/src/Hastory/Cli/OptParse.hs:
```diff
@@ -82,7 +82,7 @@ getArguments = do
 
 getEnvironment :: IO Environment
-getEnvironment = Env.parse (Env.header "hastory") (Env.prefixed "HASTORY_" envParser)
+getEnvironment = Env.parse (Env.header "hastory") envParser
```
Good, but let's also show the env vars in the `--help` screen so we can check (and for docs purposes).

NorfairKing commented:
> Please let me know how you'd like to proceed. 

It looks good to me. 
I think you're over-testing (it's the first time I tell anyone that) the option parsing. Usually I only test option parsing using regression tests if anything goes wrong.
So you can save yourself some work there

StevenXL commented:
> > Please let me know how you'd like to proceed.
> 
> It looks good to me.
> I think you're over-testing (it's the first time I tell anyone that) the option parsing. Usually I only test option parsing using regression tests if anything goes wrong.
> .. so you can save yourself some work there.

Great - thank you. I will get started on the other commands. I was having issues on how to get the [correct semantics](https://www.reddit.com/r/haskellquestions/comments/hpuomr/optparseapplicative_just_auto_vs_mayberead/) for the parsing, hence the reason why I wrote a bunch of tests until I could get them to pass. ;)

NorfairKing commented:
> I was having issues on how to get the correct semantics for the parsing, 

`optional` is the function you're looking for, from the `Alternative` class.

StevenXL commented:
> > I was having issues on how to get the correct semantics for the parsing,
> 
> `optional` is the function you're looking for, from the `Alternative` class.

👍 . Exactly this. Before I ask for another review, I am going to go over the parsing again to make sure that each method of configuring a command either error out or default consistently. 

This actually reminds me, what is the behavior that we want if a user provides an invalid value for an option that also has a default? We can 1) ignore the value that the user provided or 2) throw an error. A perfect example of this is the ListRecentDirs "bypass cache" option. In the config file, if an invalid value is provided, then the user will [see an error](https://github.com/StevenXL/hastory/blob/716df4c4ff18698d7240ee5e5adfcebd09e5da5a/hastory-cli/src/Hastory/Cli/OptParse/Types.hs#L101). However, an invalid value in the environment is [actually ignored](https://github.com/StevenXL/hastory/blob/716df4c4ff18698d7240ee5e5adfcebd09e5da5a/hastory-cli/src/Hastory/Cli/OptParse.hs#L107). (This is one the places were the behavior differs, and I'd like to standardize, but I don't know what to standardize on). Thanks for any help Syd.

StevenXL commented:
@NorfairKing I don't think it makes sense to make the `change dir` command configurable from the environment and the config file. Let me know if you disagree. If you think it should be configurable via those other two methods, then I'll change:

```haskell
newtype ChangeDirFlags =
  ChangeDirFlags
    { changeDirFlagsIdx :: Int
    }
  deriving (Show, Eq)
```

to

```haskell
newtype ChangeDirFlags =
  ChangeDirFlags
    { changeDirFlagsIdx :: Maybe Int
    }
  deriving (Show, Eq)
```

and `die` in the `combineToInstructions` action if we parse the change dir command and cannot find an index anywhere.

NorfairKing commented:
> what is the behavior that we want if a user provides an invalid value for an option that also has a default? We can 1) ignore the value that the user provided or 2) throw an error.

Definitely throw an error. All env var names are prefixed with `HASTORY`, so if any of them are set, the user _TRIED_ something, and they should see an error if they set something invalid otherwise they will not be alerted that something is wrong.

>  don't think it makes sense to make the change dir command configurable from the environment and the config file. 

I think you're right!

NorfairKing commented on hastory-cli/test/Hastory/Cli/OptParse/TypesSpec.hs:
```diff
@@ -33,8 +33,54 @@ spec =
+        let Success config = fromJSON value
+        configCacheDir config `shouldBe` Nothing
+      it "parses correctly when cache-dir key is provided and value is valid" $ do
+        value <- Data.Yaml.decodeThrow "cache-dir: ~/home"
```
This will parse as a string correctly, but `~/` is expanded by your shell so it won't mean "home" when you use it as a string like this.

StevenXL replied:
Since the user can in type that into their configuration file, and we want to do as little processing as possible, do you still want to keep this? I believe we parse this to a stronger type in `combineToInstructions` later one, using the functions in `path-io` library.

StevenXL commented:
@NorfairKing ready for another review, whenever you find some time. I did want to point out a failing test in the suite. The failing test can be found [here](https://github.com/StevenXL/hastory/blob/b3c1b2767d628153140e143fab7d9d9ca57f827e/hastory-cli/test/Hastory/Cli/OptParse/TypesSpec.hs#L46). 

This is basically testing that if you provide the `cache-dir` key in your configuration file, but you **don't** provide a value for that key, then parsing should fail since an invalid value - in this case, no value - has been provided. 

Unfortunately, when a key is provided but a corresponding value is not, the parsing library behaves as if the key/value was not provided at all, meaning that it will use the default value if one exists. This means that an invalid state has been mistaken for a valid one. I don't know if this is worth fixing, since the only invalid value for `cache-dir` is in fact an empty string; anything else is valid as far as the configuration parsing goes. 

P.S. In a related but _separate_ issue, it seems that whenever I have a default value for an option, many parsing libraries confuse the idea of "use the default value **only** if the user does not provide one" with "use this default value if you can't parse a value from the user - i.e., if the parsing fails or if the user doesn't provide the configuration". See [this issue](https://github.com/supki/envparse/pull/10) and [this one](https://github.com/supki/envparse/issues/18) for `envparse`, and [this one](https://github.com/dmjio/envy/issues/39) from `envy`. Maybe something similar is happening with `yamlparse-applicative`? It sees that the configuration that the user provided is invalid, **and** that we have a default value for this option, so it evaluates to the default value.

NorfairKing reviewed (approved):
LGTM
